### PR TITLE
IReadOnlyCollection<T> does not implement ICollection<T>

### PR DESCRIPTION
**Pull request by Pentadome**

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
@stephentoub Does the order of these checks matter?

**Comment by stephentoub** [comment]
> Does the order of these checks matter?

Just having the check itself matters. We've rejected this same change multiple times in the past, as it adds overhead to the common non-connection path and it's relatively rare to have IReadOnlyCollection types that aren't also ICollection. It's possible things have changed perfwise with @vsadov's recent casting work, but justifying this would need new numbers.

**Comment by VSadov** [comment]
Cost of checking for `ICollection<T>` may vary depending where it is in the implementation list of a given type. It is not exactly linear due to cache alignment effects and for short implementation lists, which is common, will not vary much.
Checking for `IReadOnlyCollection<T>` on common types like `List<T>` is comparable to that.
(see rough numbers in: https://github.com/dotnet/runtime/pull/1068#issuecomment-575959053)

So - yes, it is no longer a special expensive cast, but not free either. Basically just another type check that needs to be amortized by potential benefits.

File: src/libraries/System.Linq/src/System/Linq/AnyAll.cs [hunk 1]
```diff
@@ -20,6 +20,10 @@ public static bool Any<TSource>(this IEnumerable<TSource> source)
             {
                 return collectionoft.Count != 0;
             }
+            
+            if (source is IReadOnlyCollection<TSource> readOnlyCollection)
+                return readOnlyCollection.Count != 0;
```
Comment (Arrancar) [inline src/libraries/System.Linq/src/System/Linq/AnyAll.cs]: Use curly braces

**Comment by stephentoub** [comment]
There are ~20 places across the System.Linq implementation today where we check for `ICollection<T>`, a handful of which do so just to access Count.  If we want to start factoring in `IReadOnlyCollection<T>` instead of or in addition to `ICollection<T>`, I would like to see us do so holistically, meaning a) doing so consistently across the library and b) validating any performance impact to doing so (even with @VSadov's perf improvements around casting, based on measurements I took this morning, there's still a measureable impact, in particular for small-ish enumerables).  That also means considering that the a significant percentage of the collection types in use that implement `IReadOnlyCollection<T>` also implement `ICollection<T>`, so the positive benefits of such checks in many situations will be limited.  Given all that, I'm going to close this PR, which just adds a check in one place and without associated justification.  Regardless, thank you @wrkettlitz for your interest in contributing and for the PR!

**Pull request closed without merging**